### PR TITLE
Trigger skidding from impulse if the new velocity would move the unit backwards

### PR DESCRIPTION
**Pull request by loveridge**
If a unit is sufficiently knocked backwards by an impulse, I think it should cause the unit to skid backwards as an approximation of losing traction. Currently skidding is only triggered if the new velocity direction is greater than 12.9 and less than 167.1 degrees from the front direction. By maintaining the sign during the square, all new velocity vectors greater than 12.9 degrees from front will cause a skid. Impulse should also slow the unit down, but that requires a separate change.

### CODE REVIEW

File: rts/Sim/MoveTypes/GroundMoveType.h
Comment (sprunk) [inline rts/Sim/MoveTypes/GroundMoveType.h:130]: This should probably not be a macro because that can recalculate `x` twice.
It probably also shouldn't be called `SQUARE` anymore (unsure if there's a concise math term for this, maybe just "signed square"?)

Comment (loveridge) [reply rts/Sim/MoveTypes/GroundMoveType.h]: I wasn't able to find a proper mathy term either, so I called it SignedSquare as you suggested and moved it to a function.

**Comment by StanczakDominik** [comment]
Oooh, sick! Do you think this might clean up https://discord.com/channels/549281623154229250/1125040192487698453/1125040192487698453 ? I guess that probably falls under the separate change you mentioned?

**Comment by lhog** [comment]
Can you record a short video of how it looks before and after the change?

**Comment by loveridge** [comment]
> Oooh, sick! Do you think this might clean up https://discord.com/channels/549281623154229250/1125040192487698453/1125040192487698453 ? I guess that probably falls under the separate change you mentioned?

This doesn't entirely fix the issue in the discord link, since it does not cause a unit moving towards the impulse to be slowed in the cases where the impulse vector is not large enough to offset the units forward speed vector. This happens more often for units of larger mass moving at their maximum speeds.


> Can you record a short video of how it looks before and after the change?

I put some videos in the discord link above.

**Comment by marcushutchings** [comment]
Don't merge before I've had a chance to take a look at this.

**Comment by lhog** [comment]
Looks good in the video. I understand you are still looking for the ways to improve it, so not merging yet.

**Comment by loveridge** [comment]
Applying the impulse to the speed gives a slow down effect when the impulse is not strong enough to cause the skid. Here are some more comparison videos.

https://github.com/beyond-all-reason/spring/assets/13026303/e312ad82-d74e-45b0-aea4-12c0f0642c74


https://github.com/beyond-all-reason/spring/assets/13026303/9e46b961-466f-4968-ba3c-8c2bcc3f8908


https://github.com/beyond-all-reason/spring/assets/13026303/b9fed97f-aca0-4aff-9c61-e323700a0ebe


https://github.com/beyond-all-reason/spring/assets/13026303/cc15f742-49a8-4c1e-a03b-a730227b305d

**Comment by lhog** [comment]
In this batch I personally had harder time spotting the difference between the old and new.

**Comment by loveridge** [comment]
The interaction to look for is impulses applied directly opposite the units movement vector. So disregard shots that hit the unit on the side. A particularly good one is the first hit on the blitz from the 3rd pounder. You can see in the original behavior that the blitz shrugs the hit off completely and its speed is unaffected. With the proposed change the blitz speed is reduced by the impulse of the pounder's shot.
image

**Comment by marcushutchings** [comment]
@sprunk @GoogleFrog - representing other game dev teams, are you fine with this becoming standard engine behviour?

**Comment by sprunk** [comment]
How tested is this, outside of bar pounder? For example what if i have a continuous beamlaser with impulse? What if a weapon with 0.001 impulse taps something, won't this needlessly trigger skidding since [the threshold is actually 0.95 of the current speed](https://github.com/beyond-all-reason/spring/blob/08e3c9c40fe1c773c5410ef33464ed80332cbeba/rts/Sim/MoveTypes/GroundMoveType.h#L225)? What about units moving backwards.

A proper changelog entry which lists the actual effects would be nice.

Otherwise looks good.

**Comment by loveridge** [comment]
I have only tested it with BAR units which apply their impulse in single bursts. However, a unit with a fast firing impulse is not different from many units with a slow rate of fire. And it is possible to effectively "stun lock" a unit in place with the current implementation, so I think an impulse beamlaser would also keep a unit in a permanent state of skidding as long as the impulse value of the weapon is high enough.

As for small impulse numbers, the impulse velocity (`vel`) in the StartSkidding conditional is after the impulse vector is applied to the unit speed (`owner->speed + impulse`). So in order to cause a skid, a large enough impulse needs to be applied to create a resultant new speed velocity that is a certain angle different from the units `frontdir` vector (`dir` in the function). If I did my math correctly below `sqSkidSpeedMult` is only affecting the angle at which these two vectors will cause a skid, and `StartSkidding` does not account for any magnitude component. That isn't to say that the magnitudes of the impulse or unit speed don't affect skidding, just that the skidding condition is effectively a result of whether the angle between `owner->speed + impulse` and `frontdir` is greater than some angle. Which for a `sqSkidSpeedMult` of .95 is 12.92. Perhaps this variable is missnamed.

From reading the code, I believe the frontdir for units moving backwards is the direction of their movement, but I will test this when I get a chance.
I will also add an entry in the changelog.

My derivation of `Square(vel.dot(dir)) + 0.01f) < (vel.SqLength() * sqSkidSpeedMult)` (ignoring the +0.01, as it is very small effect):
math

**Comment by loveridge** [comment]
I think I also need to test this on trees, a comment on applyImpulse mentions it is how trees fall.

**Comment by loveridge** [comment]
Trees don't seem affected by the change, they fall just as wonkily as before. I did need to account for reversing direction. Also too many impulse shots could cause the unit to start moving fast enough to look weird, so I capped the induced speed at the units maximum speed + 20% extra.

File: rts/Sim/Units/Unit.cpp
Comment (sprunk) [inline rts/Sim/Units/Unit.cpp:1342]: why 1.2? wouldn't having basically no limit be fine if the unit brakes to its proper speed anyway?

**Comment by KyleAnthonyShepherd** [comment]
20% extra feels like an arbitrary enough constant, that it should be a value that can be defined somewhere gameside (either on a per-unit basis, per-impulse weapon basis, or whole game constant)

**Comment by sprunk** [comment]
A configurable limit sounds good (regardless of whether it's at the level of modrule, individual defs, or anything else) but it would be good if the default behaviour was no limit. It's up to the game dev to decide what looks silly and what doesn't.

**Comment by loveridge** [comment]
Good points! Which method of congifuration should I take, modrule unitdef, both?

**Comment by sprunk** [comment]
Both would be pointless since surely one overrides the other. Unitdef sounds better since it gives strictly more control.

File: rts/Sim/Units/Unit.cpp
Comment (marcushutchings) [reply rts/Sim/Units/Unit.cpp:1342]: Why is this multiplier here at all? What is it addressing?

**Comment by marcushutchings** [comment]
This smells like a symptom of the engine not handling physics correctly. I am concerned that putting more per-unit configuration in here runs the risk that it will likely lock the engine into more of a mess when game devs start using it and expect arbitrary behaviours to be maintained. An engine wide configuration is more palatable.

**Comment by sprunk** [comment]
Engine is indeed not handling physics correctly but mostly because there is no way for game devs to tell it what "correct" would even be. Right now all units are treated as spherical cows in a vacuum as far as things like friction or bouncing is concerned.

**Comment by loveridge** [comment]
Here was the issue I was looking to solve by capping the amount of speed an impulse can apply to a unit. I agree that just a cap is not very physics based. It does produce the effect of a sort of bump in speed if hit from behind, while not appearing too jarring. I could investigate some alternative methods, perhaps of a faster reduction to the unit's speed if above maximum in some of the speed updating functions. 


https://github.com/beyond-all-reason/spring/assets/13026303/a5094a80-3abe-4604-96e7-338bfa3c5ea1


https://github.com/beyond-all-reason/spring/assets/13026303/31f10833-f0d4-4c64-9245-730603aa3307


https://github.com/beyond-all-reason/spring/assets/13026303/795d6210-bf15-431b-b55e-dea63264765e

**Comment by KyleAnthonyShepherd** [comment]
Taking inspiration from physics.
If impulse is considered to be a flat addition to "kinetic energy" instead of "velocity", then
`impulse = Δkinetic energy = 1/2 * m * v_new^2 - 1/2 * m * v_old^2`
Remove mass for clarity:
`impulse = Δkinetic energy =v_new^2 - v_old^2`
`v_new = sqrt(impulse + v_old)`

So that might dampen the effect of multiple impulse shots landing

**Comment by GoogleFrog** [comment]
> @sprunk @GoogleFrog - representing other game dev teams, are you fine with this becoming standard engine behviour?

This looks very likely to break ZK. I'd give a 50% chance of it being fixable without further engine changes. So it is the kind of thing I would like to do when ZK is up to date and I have a weekend to mess around with it (neither of which are true at the moment). It has a significant risk of becoming a blocker, so the more configuration the better.

Consider the fact that ZK has unit-mounted impulse-only lasers that can fire at allies by design.

> This smells like a symptom of the engine not handling physics correctly. I am concerned that putting more per-unit configuration in here runs the risk that it will likely lock the engine into more of a mess when game devs start using it and expect arbitrary behaviours to be maintained. An engine wide configuration is more palatable.

The main thing to keep in mind is that the engine doesn't simply handle physics correctly or incorrectly. It can only handle physics correctly with reference to some set of goals within the design of a game, which will differ from game to game. It isn't that the current physics is intrinsically correct, it's that the current physics is essentially unconfigurable. When faced which such physics the only option was to design a game such that the answer to "what should the physics be like?" is "the only physics that is available". But that isn't to say the current physics is that far off being "right" from my perspective. It's a bit janky sometimes, but works in the important ways.

Adding configuration is a way out of this type of mess. At a minimum there should be a skid speed boost limit, and there is no reason to not make it per unit (except perhaps a technical one). Maybe a game features a bunch of impulse weapons and wants to make their hovercraft feel particularly easy to push around, so sets a high multiplier. Why preemptively close off potential?

It also sounds like the old behaviour could be retained with some sort of "skid angle" parameter (front and back?). Parameters seems fine to me provided they have a reasonably simple physical interpretation, such as a maximum speed or an impact angle that can cause skidding.

**Comment by GoogleFrog** [comment]
I am not saying let me block this. I am saying these two things:
- This PR is a risk for the engine becoming the BAR engine, by deciding which physics it wants and hardcoding it. There is no single correct physics. Configurability ameliorates this.
- It would probably take me at least a build and a day to check whether this breaks ZK. You'll be blocked if you wait for me to do this, since PRs don't generate builds. More configuration means more chances for me to tweak it before returning and saying the change is broken. Enough configuration to take it all the way to the old behaviour would be ideal for this purpose (afaik there are no bug fixes in the PR).

**Comment by KyleAnthonyShepherd** [comment]
https://github.com/beyond-all-reason/spring/blob/bcc0f71b622f5ec479c84e76bce5a829c24cc18e/rts/System/EventHandler.cpp#L445
Another thing to remember is the UnitPreDamaged callin exists, and games can choose to set `impulseMult` to zero based on their own in-game conditions, if they feel like that is necessary. So a "speed cap" can be enforced game-side with this callin, and does not need to be an engine parameter

**Comment by sprunk** [comment]
`UnitPreDamaged` has no impulse angle/dir though? So can't replicate the current behaviour with it.

**Comment by KyleAnthonyShepherd** [comment]
True, `UnitPreDamaged` would need to know about impulse direction to make fully informed decisions about `impulseMult`.

If I understand correctly, the bug being fixed is units shrugging off head on (within 12.9 degrees) impulse applications.
But there are concerns this bug is important to other games that more heavily use impulse. 

And the fix for the bug is to use the SignedSquare operation. 

Could the concerns be alleviated by a ModRule to choose between using the old Square operation to preserve old behavior, and using the new SignedSquare operation?

**Comment by GoogleFrog** [comment]
> Another thing to remember is the UnitPreDamaged callin exists

A frequent callin with a bunch of maths inside to counteract hardcoded behaviour is a poor alternative to configuring a parameter.

> Could the concerns be alleviated by a ModRule to choose between using the old Square operation to preserve old behavior, and using the new SignedSquare operation?

No. Modrules (or ideally, unitdefs) should mean something independent of some specific implementation. Configuration shouldn't merely be a toggle between some old state of the code and some new state. It has to have some physical/game design meaning. Why might people want to use either option? Independent of which happened to exist first. Otherwise we end up with the problem @marcushutchings was worried about:

> I am concerned that putting more per-unit configuration in here runs the risk that it will likely lock the engine into more of a mess when game devs start using it and expect arbitrary behaviours to be maintained.

I don't see any bugs just from reading this PR. The previous code wanted units to only skid after receiving sufficient sideways impulse. The new code additionally allows skidding when the impulse direction is backwards, and does something to reversing. I can't see a "signed square bug", it looks like the original code does what it meant to do.

Rather than add a parameter to bring back the old unsigned square, there could be parameters that set the reverse and forward thresholds required to cause a skid.

**Comment by marcushutchings** [comment]
To give this PR the best chance of success, it would be best to keep the changes small and simple. More PRs can be open to address other aspects of the topic. It would be a shame for some useful changes to be grid-locked behind one controvesial aspect that could otherwise be separated out into another PR. So, if there's a kernel of a feature we are happy with, then it would be good to push that through and then tackle to more tricky issues in another PR.

**Comment by loveridge** [comment]
I've had some time to test out these changes on Zero-K's implementation of impulse. The Jugglenaut's laser relies on causing a unit to skid. Once a unit is skidding, `UpdateSkid` properly applies the magnitude of the float3 component of `speed` to the `w` with a `SetSpeed` call which will then cause the unit to actually move faster. The weapon_impulse.lua applies an unstick impulse of 4 in the x axis which is why the following bug is present in the live version when units are pointed directly on the X axis since they will not begin skidding:

https://github.com/beyond-all-reason/spring/assets/13026303/7aa00123-c140-4b31-994b-70c21ccc2d30


With this PR, applying the impulse to the speed of the unit with SetVelocityAndSpeed has somewhat of an effect on the ability of the jugglenaut to push units in the +X direction, but it does not completely resolve the issue for some unknown reason:


https://github.com/beyond-all-reason/spring/assets/13026303/d610ead2-8eed-4b8e-880a-f8ca967a48ba


Otherwise, without the limit to the amount of impulse than can be applied to the unit's velocity, I believe Zero-K is largely unchanged. Although I am not very experienced in the current behavior. Here are some videos of it in action:


https://github.com/beyond-all-reason/spring/assets/13026303/0d7ea4fa-d6d7-44b8-9ec7-c6669631c0e0


https://github.com/beyond-all-reason/spring/assets/13026303/89a1a504-4c16-4b65-8389-c4ce529abfcf

This one with a impulse speed maximum of 4:


https://github.com/beyond-all-reason/spring/assets/13026303/4af9fb93-f98e-4a3f-940f-01663f742aab



I will be adding some configuration for the impulse speed boost maximum and the skid angle. Would it make more sense as a movedef instead of a unitdef? 

The only real bug that is being fixed is that impulse does not affect the actual movement speed of units. The other changes are somewhat additional configuration that in my opinion better approximates the expectation of an impulse being applied to vehicles. Skidding on head on impulses is to approximate a complete loss of traction. I do think it would be better with an additional threshold for how large an impulse needs to be in order to cause a skid. As it is, skids are too easy to trigger with impulses perpendicular to the units forward direction.

**Comment by GoogleFrog** [comment]
Configuration for a movedef would make no sense and be very restrictive. Unitdef is much better.

I'm not sure that impulse not affecting actual movement speed is a bug. I would have to play around with the options.The way I have been modelling impulse is that units are either in movement mode or flying mode.

**Comment by GoogleFrog** [comment]
Why is the unit moving so slowly in fun.mp4?

**Comment by loveridge** [comment]
> Why is the unit moving so slowly in fun.mp4?

In that one I set the impulse speed boost maximum to 4 times the units maximum speed.

**Comment by GoogleFrog** [comment]
Oh. There shouldn't be a speed limit when flying through the air.

**Comment by loveridge** [comment]
Speed limits were not a great solution to units moving too fast after impulses. Instead I have applied atmospheric and rolling drag to units that start moving faster than their maximum speed would allow. I have added unitdefs if games would like certain units to have different skidding and impulse receiving behaviors. As mentioned before, maybe a hovercraft would skid farther than a tank, or once boosted up to higher speeds from impulses, not slow down as fast. 

I think I would like to also apply the groundFrictionCoefficient to a unit's resistance to skidding in some fashion, but I haven't yet looked into what the logic would be. I also would like the drag from the ground to account for map gravity, but haven't gotten to that either.

Thoughts about the new speed limiting approach, and other general comments, are welcome.

**Comment by GoogleFrog** [comment]
I would like to be able to have zero atmospheric drag. Non-zero would break a few things in ZK.

**Comment by sprunk** [comment]
Unmentioned but seems to exist:
```
unitDef.rollingResistanceCoefficient = 0.05
unitDef.groundFrictionCoefficient = 0.01
```

**Comment by loveridge** [comment]
Those are for the ground, I'll add another unitdef for atmosphere.

**Comment by loveridge** [comment]
Added the atmospheric drag unitdef. You can also set the dragScales on a per unit basis with `SetSolidObjectPhysicalState`, if you perhaps have a unit that changes some sort of mode, like deploying stabilizers or something.

**Comment by lhog** [comment]
So shall we make a build and give it to @GoogleFrog for testing?

**Comment by loveridge** [comment]
I'm happy with how it turned out. Other features like map gravity affecting skids will be a new MR.

**Comment by lhog** [comment]
Here's the Windows binary [spring.zip](https://github.com/beyond-all-reason/spring/files/12553466/spring.zip)
It should be compatible with the base of [latest pre-release](https://github.com/beyond-all-reason/spring/releases/tag/spring_bar_%7BBAR105%7D105.1.1-1923-gbd815da).

**Comment by lhog** [comment]
@GoogleFrog did you test the binary?

**Comment by GoogleFrog** [comment]
Not yet. September has been busy and October is looking busy too. Perhaps next week.

**Comment by StanczakDominik** [comment]
Sorry to bother, but I guess it's not looking good for a test in October, @GoogleFrog?

**Comment by GoogleFrog** [comment]
spring_cj1rK4UvPn
I made a unit bounce forever by changing the UNSTICK_CONSTANT to point in the y direction in the impulse gadget, then shooting a unit with Jugglenaut.

**Comment by GoogleFrog** [comment]
I was able to make unlimited bouncing just with a tall cliff and mundane impulse.

**Comment by loveridge** [comment]
> I made a unit  bounce forever by changing the UNSTICK_CONSTANT to point in the y direction in the impulse gadget, then shooting a unit with Jugglenaut.

Negative Y impulse has no effect while the unit has the onground state, so your impulse gadget is adding a positive Y unstick_constant and not subtracting it. I believe this is to prevent the unit from being knocked below the surface. 

weapon_impulse.lua currently triggers a skid using the X axis, it could instead trigger a skid perpendicular to the frontdir if you are trying to fix the unit not skidding when it is facing only X axis:
```lua
if data.useDummy then
    local dx, dy, dz = spGetUnitDirection(unitID)
    spAddUnitImpulse(unitID, -dz * UNSTICK_CONSTANT, 0, dx * UNSTICK_CONSTANT) --dummy impulse (applying perpendicular to frontdir causes a skid)
    spAddUnitImpulse(unitID, data.x, data.y, data.z)
    spAddUnitImpulse(unitID, dz * UNSTICK_CONSTANT, 0, -dx * UNSTICK_CONSTANT) --remove dummy impulse
else
```

**Comment by GoogleFrog** [comment]
> I was able to make unlimited bouncing just with a tall cliff and mundane impulse. It goes away with luarules disable though.

Ie, the UNSTICK_CONSTANT part of my bouncing unit comment was irrelevant.

**Comment by GoogleFrog** [comment]
On UNSTICK_CONSTANT, could there be a callin to unstick, rather than doing the add-then-subtract hax? Since we're on the engine repo it feels more appropriate to talk about how to remove a hax, than how to fix a bug with one.

**Comment by loveridge** [comment]
> > I was able to make unlimited bouncing just with a tall cliff and mundane impulse. It goes away with luarules disable though.
> 
> Ie, the UNSTICK_CONSTANT part of my bouncing unit comment was irrelevant.

I'm not able to reproduce this without a Y axis unstick_constant. I've tried knocking a unit off a cliff with a jugglenaut, newton, and an archer, and they do not continue to bounce. Do you have more steps to reproduce?

**Comment by loveridge** [comment]
> On UNSTICK_CONSTANT, could there be a callin to unstick, rather than doing the add-then-subtract hax? Since we're on the engine repo it feels more appropriate to talk about how to remove a hax, than how to fix a bug with one.

`SetPhysicalStateBit` could be exposed via lua, and then you would be able to set the unit as skidding or flying.

**Comment by GoogleFrog** [comment]
[2023-10-17_08-55-42-607_FolsomDamDeluxeV4_105.1.1-2031-gd4526bf BAR105.sdfz.txt](https://github.com/beyond-all-reason/spring/files/13278414/2023-10-17_08-55-42-607_FolsomDamDeluxeV4_105.1.1-2031-gd4526bf.BAR105.sdfz.txt)

Dunno, I just shot something off a cliff and they bounced. But not in this replay. Maybe something has to bounce before everything can. Other weird things happened in this replay though, like a tank hitting the edge of a wall and flying across the map.

**Comment by loveridge** [comment]
> [2023-10-17_08-55-42-607_FolsomDamDeluxeV4_105.1.1-2031-gd4526bf BAR105.sdfz.txt](https://github.com/beyond-all-reason/spring/files/13278414/2023-10-17_08-55-42-607_FolsomDamDeluxeV4_105.1.1-2031-gd4526bf.BAR105.sdfz.txt)
> 
> Dunno, I just shot something off a cliff and they bounced. But not in this replay. Maybe something has to bounce before everything can. Other weird things happened in this replay though, like a tank hitting the edge of a wall and flying across the map.

Is this the replay you were thinking of? It seems like just 2 hours of building solar collectors:
image

**Comment by GoogleFrog** [comment]
Sorry, this one.
[2023-11-07_09-49-36-999_FolsomDamDeluxeV4_105.1.1-1982-g5a72551 feature_skid-180-impulse.sdfz.txt](https://github.com/beyond-all-reason/spring/files/13296291/2023-11-07_09-49-36-999_FolsomDamDeluxeV4_105.1.1-1982-g5a72551.feature_skid-180-impulse.sdfz.txt)

I see the replay format changed, and I got the previous one via sort by name.

**Comment by sprunk** [comment]
> it's just 2 hours of building solar collectors

gigachad.jpg

**Comment by loveridge** [comment]
> Sorry, this one. [2023-11-07_09-49-36-999_FolsomDamDeluxeV4_105.1.1-1982-g5a72551 feature_skid-180-impulse.sdfz.txt](https://github.com/beyond-all-reason/spring/files/13296291/2023-11-07_09-49-36-999_FolsomDamDeluxeV4_105.1.1-1982-g5a72551.feature_skid-180-impulse.sdfz.txt)
> 
> I see the replay format changed, and I got the previous one via sort by name.

Just remove the upMag from your new impulse gadget changes.

**Comment by sprunk** [comment]
In the video above, the gravity turrets didn't actually push anything. upMag is there so they detach from the ground which removes "ground friction" (which may not be modelled as such in engine but it is how it looks like). Won't keeping only the XZ components result in just drifting without sending the unit flying? That would be an unwanted behaviour change.

**Comment by loveridge** [comment]
I think I had a bad merge with the impulse gadget. But there was still some wonkiness, probably due to the large unstick impulses not accounting for the ground normal. So it is much easier to just add the lua call to set your own physical state bit. Here is the resultant weapon_impulse gadget:
```lua
if data.useDummy then
  Spring.SetUnitPhysicalStateBit(unitID, 512) -- skidding
  Spring.SetUnitPhysicalStateBit(unitID, 128) -- flying
  spAddUnitImpulse(unitID, data.x, data.y, data.z)
else
```

**Comment by loveridge** [comment]
The video of the demo:

https://github.com/beyond-all-reason/spring/assets/13026303/18ad9491-535e-408c-a55e-5941af390bac

**Comment by sprunk** [comment]
when using the "set skidding" call, can 1 gravity turret push a `striderdetriment` or prevent it from moving? one of the good effects of using raw impulse to unstick units is that massive units are resistant to this somewhat.

**Comment by loveridge** [comment]
This is how the game currently works, so it is hard for me to know what is supposed to be happening

https://github.com/beyond-all-reason/spring/assets/13026303/99e2e8a0-9eaf-4246-ae0a-17abdd537c11

**Comment by GoogleFrog** [comment]
On zk 8afbfaa072473a1298b24e9dbb1b847b56cdb0b9 the third Minotaur flies across the map after sliding down the wall for a bit.
image

**Comment by GoogleFrog** [comment]
It looks like it was a ZK bug, but I only found it after two weeks of testing. After fiddling around with it, I think this PR is similar, edge cases will be rare so are only really going to be found through volume. That means the most important thing is for someone to be around to fix it when it is live tested, and being able to configure it improves things as it shortens the response time for live testing (since game devs can deploy tests rather than wait for an engine build).

But anyway, I said don't necessarily make a blocker ages ago, and now it looks like if I am, then this will probably be blocked forever. So I shouldn't be. Because this isn't the sort of thing I can test like this. I hit the limit of what I can test without volume and it didn't obviously not work. So we just have to be ready to fix it if/when it turns out something broke.

**Comment by loveridge** [comment]
@lhog When you have some time can you give GoogleFrog another binary with the new change that will let him test setting skidding directly with `Spring.SetUnitPhysicalStateBit`

**Comment by lhog** [comment]
Sure, don't mind if I rebase?

**Comment by loveridge** [comment]
Sure, I don't mind.

**Comment by lhog** [comment]
@loveridge I don't think I can push easily to your repo. Can you rebase yourself and I'll make the exe.

**Comment by loveridge** [comment]
Now up to date with BAR105.

**Comment by lhog** [comment]
@GoogleFrog the engine build for your tests, use [the ](https://github.com/beyond-all-reason/spring/releases/tag/spring_bar_%7BBAR105%7D105.1.1-2110-g4903e7a) as a base.

[spring.zip](https://github.com/beyond-all-reason/spring/files/13404737/spring.zip)

**Review by sprunk — APPROVED** [review]

**Comment by GoogleFrog** [comment]
Am I supposed to click something to approve this? I already said. I don't want me approving this to be a requirement. It might break something, but we won't know until it is actually used. The important thing is to have support for this after it is merged.

> But anyway, I said don't necessarily make a blocker ages ago, and now it looks like if I am, then this will probably be blocked forever. So I shouldn't be. Because this isn't the sort of thing I can test like this. I hit the limit of what I can test without volume and it didn't obviously not work. So we just have to be ready to fix it if/when it turns out something broke.

**Comment by lhog** [comment]
Let's merge when the main branch is unfrozen and see how it goes, if anything we can always revert, but I expect this change to be well maintained.

**Comment by marcushutchings** [comment]
Argeed. Testing has been done on best endeavours, which is as good as we can hope for in order to derisk the release.

**Comment by lhog** [comment]
I'm squashing this change into one commit and pushing it to the mainline, the feature branch will remain intact in case @loveridge or someone else would like iterate on the past commits.

Sorry for the long waiting period.

**Pull request merged**
